### PR TITLE
Displaying folder-based permissions in UI rather than file-based

### PR DESCRIPTION
**Pull request by Taloth**
#### Database Migration
**YES** = 145

#### Description
1. Swapped the file-based permissions (644) with folder-based permissions (755). It's technically exactly the same but it was confusing for people.
2. Also extended EnhancedSelectInput with a editable mode, allowing uses to type in whatever they want but still have a dropdown with common presets.
I'm not entirely happy with it, I think that maybe duplicating the dropdown code to a PresetTextInput might be better. So let me know what you think.
3. Added preview of umask, folder and file permissions including drwxr-xr-x notation.
4. Readded chgrp, there was an usecase when wanting to secure Sonarr db vs the media library.
5. Fixed bug that erroneously removed existing setgid flags when applying permissions.

image

#### Todos
- [x] Cleanup presets
- [ ] Tests for setgid handling
- [x] Some known UI React errors that need to be fixed

#### Issues Fixed or Closed by this PR

*

### CODE REVIEW

**Comment by Taloth** [comment]
@markus101 Can you look this over? As mentioned I'm not entirely happy with the UI changes and perhaps there's a better way to do that.

**Comment by RobinDadswell** [comment]
Just my thoughts on this but is there a way to make it more obvious that you can type and put whatever octet you want in as well, not that I can think of any others that you would want but you never know!

File: frontend/src/Components/Form/UMaskInput.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    result = digit + result;
+  }
+
+  return result;
+}
+
+class UMaskInput extends Component {
+
+  //
+  // Lifecycle
```
Comment (markus101) [inline frontend/src/Components/Form/UMaskInput.js:64]: Nit: clean up the empty blocks

File: frontend/src/Settings/MediaManagement/MediaManagement.js [hunk 2]
```diff
@@ -382,17 +382,34 @@ class MediaManagement extends Component {
...
                         isAdvanced={true}
                       >
-                        <FormLabel>File chmod mode</FormLabel>
+                        <FormLabel>chmod Folder</FormLabel>
+
+                        <FormInputGroup
+                          type={inputTypes.UMASK}
+                          name="chmodFolder"
+                          helpText="Octal, applied during import/rename to media folders and files (without execute bits)"
+                          helpTextWarning="This only works if the user running sonarr is owner of the file. It's better to ensure the download client sets the permissions properly."
```
Comment (markus101) [inline frontend/src/Settings/MediaManagement/MediaManagement.js:391]: `...user running Sonarr is the owner...`

File: frontend/src/Settings/MediaManagement/MediaManagement.js [hunk 3]
```diff
@@ -382,17 +382,34 @@ class MediaManagement extends Component {
...
                         <FormInputGroup
                           type={inputTypes.TEXT}
-                          name="fileChmod"
-                          helpTexts={[
-                            'Octal, applied to media files when imported/renamed by Sonarr',
-                            'The same mode is applied to series/season folders with the execute bit added, e.g., 0644 becomes 0755'
-                          ]}
+                          name="chownGroup"
+                          helpText="Group name or gid. Use gid for remote file systems."
+                          helpTextWarning="This only works if the user running sonarr is owner of the file. It's better to ensure the download client uses the same group as sonarr."
```
Comment (markus101) [inline frontend/src/Settings/MediaManagement/MediaManagement.js:408]: `...user running Sonarr is the owner...`

File: frontend/src/Settings/MediaManagement/MediaManagement.js [hunk 4]
```diff
@@ -382,17 +382,34 @@ class MediaManagement extends Component {
...
                       >
-                        <FormLabel>File chmod mode</FormLabel>
+                        <FormLabel>chmod Folder</FormLabel>
+
+                        <FormInputGroup
+                          type={inputTypes.UMASK}
+                          name="chmodFolder"
+                          helpText="Octal, applied during import/rename to media folders and files (without execute bits)"
+                          helpTextWarning="This only works if the user running sonarr is owner of the file. It's better to ensure the download client sets the permissions properly."
+                          // helpLink="https://github.com/sonarr/sonarr/wiki#asdf"
```
Comment (markus101) [inline frontend/src/Settings/MediaManagement/MediaManagement.js:392]: Remove this and the one below if we're not including docs on how this works

File: src/NzbDrone.Core/Datastore/Migration/145_swap_filechmod_for_folderchmod.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                getFileChmodCmd.Transaction = tran;
+                getFileChmodCmd.CommandText = @"SELECT Value FROM Config WHERE Key = 'filechmod'";
+                
+                var fileChmod = getFileChmodCmd.ExecuteScalar() as string;
+                if (fileChmod != null)
+                {
+                    if (fileChmod.IsNotNullOrWhiteSpace())
+                    {
+                        // Convert with touching mono libraries. We take the 'r' bits and shifting them to the 'x' position, preserving everything else.
```
Comment (markus101) [inline src/NzbDrone.Core/Datastore/Migration/145_swap_filechmod_for_folderchmod.cs:30]: `Convert without touching`?

**Review by markus101 — APPROVED** [review]
To Robin's point, it's a bit hidden that you can type your own, but not really sure how to better indicate it is (and we do use similar things elsewhere).

Mostly just nits, I think the UI works really well and besides the aforementioned bit, but I see no reason to hold things up trying to improve anything.

**Pull request merged**